### PR TITLE
Step 9: Audience

### PR DESCRIPTION
**Pull request by divbyzero**
- Added EvolvAudienceQuery
- Added optional audienceQuery to EvolvRawAllocation
- Added isFilter to EvolvRawAllocation
- Updated EvolvAllocations (added audience)
- Updated EvolvExecution
- Updated EvolvEventEmitter (fixed issues with debug logging)
- Updated AllocationsTest
- Added AudienceTest

### CODE REVIEW

File: Source/EvolvAudienceQuery.swift [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    }
+    
+    static func evaluate(_ expression: EvolvQuery, userAttributes: [String: String]) -> Bool {
+        switch expression {
+        case .rule(let rule):
+            switch (rule.operator, rule.value) {
+            case (.exists, let value as String):
+                return userAttributes.keys.contains(where: { $0 == value })
+            case (.equal, let values as [String]) where values.count > 1:
+                return userAttributes.keys.contains(where: { $0 == values[0] }) && userAttributes[values[0]] == values[1]
```
Comment (phylliswong) [inline Source/EvolvAudienceQuery.swift:106]: can we add a doc string for this method?

File: Source/EvolvRawAllocation.swift [hunk 2]
```diff
@@ -34,6 +34,8 @@ public class EvolvRawAllocation: NSObject, Decodable {
     public let genome: EvolvRawAllocationNode
     // TODO: add doc for property
     public let excluded: Bool
+    // TODO: add doc for property
+    public let audienceQuery: EvolvQuery?
```
Comment (phylliswong) [inline Source/EvolvRawAllocation.swift:38]: #
```suggestion
   /// Query the specific audience filter defined by project
    public let audienceQuery: EvolvQuery?
```

File: Source/EvolvRawAllocation.swift [hunk 3]
```diff
@@ -34,6 +34,8 @@ public class EvolvRawAllocation: NSObject, Decodable {
     public let genome: EvolvRawAllocationNode
     // TODO: add doc for property
     public let excluded: Bool
```
Comment (phylliswong) [inline Source/EvolvRawAllocation.swift:36]:
```suggestion
    /// User has been excluded from the experiment
    public let excluded: Bool
```

File: Source/EvolvRawAllocation.swift [hunk 4]
```diff
@@ -34,6 +34,8 @@ public class EvolvRawAllocation: NSObject, Decodable {
     public let genome: EvolvRawAllocationNode
```
Comment (phylliswong) [inline Source/EvolvRawAllocation.swift:34]:
```suggestion
    /// The unique identifier for the specific permutations for the variant values 
    public let candidateId: String
```

```suggestion
    /// The specific treatments to be applied to the end users experience
    public let genome: EvolvRawAllocationNode
```

**Review by phylliswong — APPROVED** [review]
Non-blocking: suggested a few things to do with doc strings. Added some definitions that were labeled TODO. Everything looks great, thanks for doing this.

File: Source/EvolvAudienceQuery.swift (hunk 1 above)
Comment (divbyzero) [reply Source/EvolvAudienceQuery.swift:106]: /// Recursive handling of the query to check if a certain rule will work
    /// for the current userAttributes
    ///
    /// - Parameters:
    ///   - expression: Query the specific audience filter defined by project
    ///   - userAttributes: map representing attributes that represent the participant
    /// - Returns: Returns true if rule is matched with userAttributes

File: Source/EvolvAudienceQuery.swift [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        if let rule = try? container.decode(Rule.self) {
+            self = .rule(rule)
+        } else if let compoundRule = try? container.decode(CompoundRule.self) {
+            self = .compoundRule(compoundRule)
+        } else {
+            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Mismatched Types")
+        }
+    }
+    
+    static func evaluate(_ expression: EvolvQuery, userAttributes: [String: String]) -> Bool {
```
Comment (divbyzero) [inline Source/EvolvAudienceQuery.swift:99]:
```suggestion
    /// Recursive handling of the query to check if a certain rule will work
    /// for the current userAttributes
    ///
    /// - Parameters:
    ///   - expression: Query the specific audience filter defined by project
    ///   - userAttributes: map representing attributes that represent the participant
    /// - Returns: Returns true if rule is matched with userAttributes
    static func evaluate(_ expression: EvolvQuery, userAttributes: [String: String]) -> Bool {
```

File: Source/EvolvAudienceQuery.swift (hunk 1 above)
Comment (phylliswong) [reply Source/EvolvAudienceQuery.swift:106]: this is great

File: Source/EvolvAudienceQuery.swift (hunk 5 above)
Comment (phylliswong) [reply Source/EvolvAudienceQuery.swift:99]: this works, if you want to add this to the PR along with the other suggestions, then I can merge.

**Pull request merged**